### PR TITLE
Set compatibility config to NONE for protobuf subjects

### PR DESCRIPTION
**Pull request by siddhant2001**

### CODE REVIEW

**Comment by ravisingal** [comment]
@kishansairam9 is working on a  generic solution.

**Comment by kishansairam9** [comment]
> @kishansairam9 is working on a generic solution.

Adding each proto is painful, there has been a meeting where laxman and team decided on having a cron job to set for all protobufs. No need to override each proto individually. I think this is the way to go forward, my old PR is abandoned

File: compatibility-changer/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,22 @@
+FROM python:3.12.4-slim
+
+# update packages
+RUN apt update && apt upgrade -y
+
+RUN apt install -y curl && \
```
Comment (laxmanchekka) [inline compatibility-changer/Dockerfile:6]: Do we need them?

File: compatibility-changer/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    apt install -y wget
+
+# setup virtual environment
+RUN python -m venv /opt/venv
+
+# create directory and copy pinot-data-metrics directory
```
Comment (laxmanchekka) [inline compatibility-changer/Dockerfile:12]: Fix the comments

File: compatibility-changer/compatibility-changer/requirements.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+certifi==2024.8.30
```
Comment (laxmanchekka) [inline compatibility-changer/compatibility-changer/requirements.txt:1]: do we need all of them explicitly? should we omit transitive dependencies and let pip figure it out?

File: helm/templates/compatibility-changer/compatibility-changer-configmap.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ .Values.compatibilityChanger.config.name }}
+  labels:
+    release: {{ .Release.Name }}
+data:
+  application.conf: |
+    {
+      "defaultSchemaType": {{ .Values.compatibilityChanger.config.defaultSchemaType }}
```
Comment (laxmanchekka) [inline helm/templates/compatibility-changer/compatibility-changer-configmap.yaml:11]: schemaType -> compatibilityType

File: compatibility-changer/compatibility-changer/cron-job.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+import logging
+import os
+import urllib.parse
+from tqdm import tqdm
+
+# Setup logging
+logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
+
+# Read helm values
+host = os.getenv('SCHEMA_REGISTRY_HOST', 'localhost')
```
Comment (laxmanchekka) [inline compatibility-changer/compatibility-changer/cron-job.py:11]: make this configurable

File: compatibility-changer/compatibility-changer/cron-job.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    response = requests.put(url, json={"compatibilityLevel": "NONE"})
+    logging.info(f"PUT {url} response: {response.json}")
+
+def main():
+    subjects = get_subjects()
+    logging.info(f"All subjects: {subjects}")
+    for subject in tqdm(subjects):
+        versions = get_subject_versions(subject)
+        if not versions:
+            continue
```
Comment (laxmanchekka) [inline compatibility-changer/compatibility-changer/cron-job.py:54]: log a warn when no version available.

File: compatibility-changer/compatibility-changer/cron-job.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+def set_compatibility_level(subject):
+    encoded_subject = url_encode(subject)
+    url = f"{base_url}/config/{encoded_subject}"
+    logging.info(f"Setting compatibility level to NONE for subject: {subject} with URL: {url}")
+    response = requests.put(url, json={"compatibilityLevel": "NONE"})
+    logging.info(f"PUT {url} response: {response.json}")
+
+def main():
+    subjects = get_subjects()
+    logging.info(f"All subjects: {subjects}")
```
Comment (laxmanchekka) [inline compatibility-changer/compatibility-changer/cron-job.py:50]: debug log or just log number of subjects

File: compatibility-changer/compatibility-changer/cron-job.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    response = requests.get(url)
+    return response.json()
+
+def get_subject_versions(subject):
+    encoded_subject = url_encode(subject)
+    url = f"{base_url}/subjects/{encoded_subject}/versions"
+    response = requests.get(url)
+    return response.json()
+
+def get_subject_version_details(subject, version):
```
Comment (laxmanchekka) [inline compatibility-changer/compatibility-changer/cron-job.py:29]: get_subject_version_details -> get_schema_version

File: compatibility-changer/compatibility-changer/cron-job.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    subjects = get_subjects()
+    logging.info(f"All subjects: {subjects}")
+    for subject in tqdm(subjects):
+        versions = get_subject_versions(subject)
+        if not versions:
+            continue
+
+        # Select the first version
+        first_version = versions[0]
+        details = get_subject_version_details(subject, first_version)
```
Comment (laxmanchekka) [inline compatibility-changer/compatibility-changer/cron-job.py:58]: details -> schema

File: compatibility-changer/compatibility-changer/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+charset-normalizer==3.3.2
+idna==3.10
+requests==2.32.3
+tqdm==4.66.5
```
Comment (laxmanchekka) [inline compatibility-changer/compatibility-changer/requirements.txt:5]: Comment from @ravisingal : Progress bar may not be required for cron.

**Review by laxmanchekka — CHANGES_REQUESTED** [review]

**Review by laxmanchekka — DISMISSED** [review]

**Comment by laxmanchekka** [comment]
@ravisingal : can you please check this from helm, k8s and git actions perspective as well

File: helm/templates/schema-compatibility/cronjob.yaml [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+          imagePullSecrets:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
+          restartPolicy: {{ .Values.schemaCompatibility.cronjob.restartPolicy }}
+          affinity:
+            {{- toYaml .Values.schemaCompatibility.cronjob.affinity | nindent 12 }}
+          nodeSelector:
+            {{- toYaml .Values.schemaCompatibility.cronjob.nodeSelector | nindent 12 }}
+          securityContext:
+            {{- toYaml .Values.schemaCompatibility.cronjob.securityContext | nindent 12 }}
```
Comment (ravisingal) [inline helm/templates/schema-compatibility/cronjob.yaml:60]:
```suggestion
            {{- toYaml .Values.podSecurityContext | nindent 12 }}
```

File: helm/templates/schema-compatibility/cronjob.yaml [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+                {{- toYaml . | nindent 14 }}
+              {{- end }}
+              volumeMounts:
+                - name: service-config
+                  mountPath: /app/resources/configs/schema-compatibility/application.conf
+                  subPath: application.conf
+          {{- with .Values.imagePullSecrets }}
+          imagePullSecrets:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
```
Comment (ravisingal) [inline helm/templates/schema-compatibility/cronjob.yaml:53]:
```suggestion
          {{- end }}
          {{- with .Values.containerSecurityContext }}
          securityContext:
            {{- toYaml . | nindent 14 }}
          {{- end }}
```

File: helm/templates/schema-compatibility/cronjob.yaml [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+  schedule: {{ .Values.schemaCompatibility.cronjob.schedule }}
+  startingDeadlineSeconds: {{ .Values.schemaCompatibility.cronjob.startingDeadlineSeconds }}
+  successfulJobsHistoryLimit: {{ .Values.schemaCompatibility.cronjob.successfulJobsHistoryLimit }}
+  suspend: {{ .Values.schemaCompatibility.cronjob.suspend }}
+  jobTemplate:
+    spec:
+      activeDeadlineSeconds: {{ .Values.schemaCompatibility.cronjob.activeDeadlineSeconds }}
+      template:
+        metadata:
+          {{- with .Values.schemaCompatibility.cronjob.podAnnotations }}
```
Comment (ravisingal) [inline helm/templates/schema-compatibility/cronjob.yaml:19]:
```suggestion
          {{- with merge .Values.schemaCompatibility.cronjob.podAnnotations .Values.commonPodLabels }}
```

File: helm/templates/schema-compatibility/cronjob.yaml [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+          annotations:
+            {{- toYaml . | nindent 12 }}
+          {{- end }}
+          {{- with .Values.schemaCompatibility.cronjob.podLabels }}
```
Comment (ravisingal) [inline helm/templates/schema-compatibility/cronjob.yaml:23]:
```suggestion
          {{- with merge .Values.schemaCompatibility.cronjob.podLabels .Values.commonPodLabels }}
```

File: helm/values.yaml [hunk 15]
```diff
@@ -146,6 +146,39 @@ servicemonitor:
...
+    podAnnotations: {}
+    schedule: "0 * * * *" # Hourly.
+    suspend: false
+    successfulJobsHistoryLimit: 3
+    resources: {}
+    concurrencyPolicy: Forbid
+    restartPolicy: Never
+    affinity: {}
+    nodeSelector: {}
+    securityContext: {}
```
Comment (ravisingal) [inline helm/values.yaml:177]:
```suggestion
```

File: schema-compatibility/build.gradle.kts [hunk 16]
```diff
@@ -0,0 +1,11 @@
+plugins {
+  id("base")
+  id("org.hypertrace.docker-plugin")
+  id("org.hypertrace.docker-publish-plugin")
+}
+
+hypertraceDocker {
+  defaultImage {
+    imageName.set("schema-compatibility")
+  }
```
Comment (kishansairam9) [inline schema-compatibility/build.gradle.kts:10]: add `dockerFile.set(file("Dockerfile"))`

File: schema-compatibility/build.gradle.kts [hunk 17]
```diff
@@ -0,0 +1,11 @@
+plugins {
+  id("base")
+  id("org.hypertrace.docker-plugin")
+  id("org.hypertrace.docker-publish-plugin")
+}
+
+hypertraceDocker {
+  defaultImage {
+    imageName.set("schema-compatibility")
```
Comment (kishansairam9) [inline schema-compatibility/build.gradle.kts:9]: do we need this?

**Review by laxmanchekka — APPROVED** [review]

**Pull request merged**